Lib/test/test_httpservers.py

```diff
@@ -1535,6 +1537,197 @@ def test_server_test_ipv4(self, _):
+            'protocol': self.default_protocol,
+            'port': self.default_port,
+            'bind': self.default_bind,
+            'tls_cert': None,
+            'tls_key': None,
+            'tls_password': None,
+        }
+        return super().setUp()
+
+    def tearDown(self):
+        if os.path.exists(self.tls_password_file):
+            os.remove(self.tls_password_file)
+        return super().tearDown()
+
+    def get_random_temporary_file_name(self):
+            return os.path.join(tempfile.gettempdir(), os.urandom(16).hex())
+
+    def invoke_httpd(self, *args):
+        output = StringIO()
+        with contextlib.redirect_stdout(output), \
+            contextlib.redirect_stderr(output):
+            server._main(args)
+        return textwrap.dedent(output.getvalue()).strip()
+
+
+    def test_port_flag(self, mock_func):
+        ports = [8000, 65535,]
+        for port in ports:
+            with self.subTest(port=port):
+                self.invoke_httpd(str(port))
+                self.args['port'] = port
+                mock_func.assert_called_once_with(**self.args)
+                self.args['port'] = self.default_port
+                mock_func.reset_mock()
+
+    def test_directory_flag(self, mock_func):
+        options = ['-d', '--directory']
+        directories = ['.', '/foo', '\\bar', '/',
+                       'C:\\', 'C:\\foo', 'C:\\bar',]
+        for flag in options:
+            for directory in directories:
+                with self.subTest(flag=flag, directory=directory):
+                    self.invoke_httpd(flag, directory)
+                    mock_func.assert_called_once_with(**self.args)
+                    mock_func.reset_mock()
+
+    def test_bind_flag(self, mock_func):
+        options = ['-b', '--bind']
+        bind_addresses = ['localhost', '127.0.0.1', '::1',
+                          '0.0.0.0', '8.8.8.8',]
+        for flag in options:
+            for bind_address in bind_addresses:
+                with self.subTest(flag=flag, bind_address=bind_address):
+                    self.invoke_httpd(flag, bind_address)
+                    self.args['bind'] = bind_address
+                    mock_func.assert_called_once_with(**self.args)
+                    self.args['bind'] = self.default_bind
+                    mock_func.reset_mock()
+
+    def test_protocol_flag(self, mock_func):
+        options = ['-p', '--protocol']
+        protocols = ['HTTP/1.0', 'HTTP/1.1', 'HTTP/2.0', 'HTTP/3.0',]
+        for flag in options:
+            for protocol in protocols:
+                with self.subTest(flag=flag, protocol=protocol):
+                    self.invoke_httpd(flag, protocol)
+                    self.args['protocol'] = protocol
+                    mock_func.assert_called_once_with(**self.args)
+                    self.args['protocol'] = self.default_protocol
+                    mock_func.reset_mock()
+
+    def test_cgi_flag(self, mock_func):
```

Comment:
Add a TODO saying that this should be removed once the CGI component is removed as well.